### PR TITLE
x86_cpu_protection_key: support protection keys

### PR DESCRIPTION
**Pull request by nanliu-r**
ID: 2094628

New case which support protection keys in an
AMD EPYC-Milan or newer models VM.

Signed-off-by: nanliu <nanliu@redhat.com>

### CODE REVIEW

**Comment by nanliu-r** [comment]
(1/1) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.1.0.x86_64.io-github-autotest-qemu.x86_cpu_protection_key.q35: PASS (64.78 s)

@PaulYuuu Please help to review this, thanks.

File: qemu/tests/x86_cpu_protection_key.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    params["start_vm"] = "yes"
+    vm_name = params['main_vm']
+    env_process.preprocess_vm(test, params, env, vm_name)
+
+    vm = env.get_vm(vm_name)
+    error_context.context("Try to log into guest", test.log.info)
+    session = vm.wait_for_login()
+
+    guest_dir = params["guest_dir"]
+    timeout = int(params["timeout"])
```
Comment (PaulYuuu) [inline qemu/tests/x86_cpu_protection_key.py:36]:
```suggestion
    timeout = params.get_numeric("timeout")
```

File: qemu/tests/x86_cpu_protection_key.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    kernel_v = session.cmd_output("uname -r").strip()
+    mkdir_cmd = session.cmd('mkdir -p %s' % guest_dir)
+    src_rpm = "kernel-" + kernel_v[:-7] + ".src.rpm"
```
Comment (PaulYuuu) [inline qemu/tests/x86_cpu_protection_key.py:39]: `kernel_v[:-7]` could not always return the correct version. The string length of different architectures is not the same. Suggest use `.rsplit(".", 1)`

File: qemu/tests/x86_cpu_protection_key.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    linux_name = "linux-" + kernel_v[:-7]
+    download_rpm_cmd = 'cd %s &&' % guest_dir + params["download_rpm_cmd"] % src_rpm
```
Comment (PaulYuuu) [inline qemu/tests/x86_cpu_protection_key.py:41]: Would better add a space " " after `&&`

**Comment by nanliu-r** [comment]
@PaulYuuu Hi, I limit the guest versions since the feature doesn't be supported in these guests.
Please help to review again, thanks.

(1/1) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.1.0.x86_64.io-github-autotest-qemu.x86_cpu_protection_key.q35: PASS (340.04 s)

File: qemu/tests/x86_cpu_protection_key.py (hunk 2 above)
Comment (nanliu-r) [reply qemu/tests/x86_cpu_protection_key.py]: > `kernel_v[:-7]` could not always return the correct version. The string length of different architectures is not the same. Suggest use `.rsplit(".", 1)`

updated and thanks.

File: qemu/tests/x86_cpu_protection_key.py (hunk 1 above)
Comment (nanliu-r) [reply qemu/tests/x86_cpu_protection_key.py]: > 

updated and thanks.

File: qemu/tests/x86_cpu_protection_key.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    4) Check results
+
+    :param test:   QEMU test object.
+    :param params: Dictionary with the test parameters.
+    :param env:    Dictionary with test environment.
+    """
+    unsupported_models = params.get("unsupported_models", "")
+    cpu_model = params.get("cpu_model")
+    if not cpu_model:
+        cpu_model = cpu.get_qemu_best_cpu_model(params)
```
Comment (PaulYuuu) [inline qemu/tests/x86_cpu_protection_key.py:23]:
```suggestion
    cpu_model = params.get("cpu_model", cpu.get_qemu_best_cpu_model(params))
```

File: qemu/tests/cfg/x86_cpu_protection_key.cfg [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    required_qemu = [6.2, )
+    start_vm = no
+    timeout = 120
+    cpu_model_flags += ",enforce"
+    unsupported_models = [EPYC-Rome, EPYC, Opteron_G5, Opteron_G4, Opteron_G3, Opteron_G2, Opteron_G1]
+    guest_dir = '/home/tpm-test/'
+    test_dir = '/tools/testing/selftests/vm/'
+    download_rpm_cmd = brew download-build --rpm %s
+    uncompress_cmd_src = rpm2cpio *.src.rpm |cpio -iv
+    uncompress_cmd = tar -xvf *.tar.xz
```
Comment (PaulYuuu) [inline qemu/tests/cfg/x86_cpu_protection_key.cfg:16]:
```
# tar --help
-J, --xz                   filter the archive through xz
```

File: qemu/tests/x86_cpu_protection_key.py (hunk 4 above)
Comment (nanliu-r) [reply qemu/tests/x86_cpu_protection_key.py]: update and thanks.

File: qemu/tests/cfg/x86_cpu_protection_key.cfg (hunk 5 above)
Comment (nanliu-r) [reply qemu/tests/cfg/x86_cpu_protection_key.cfg]: > ```
> # tar --help
> -J, --xz                   filter the archive through xz
> ```

update and thanks.

**Comment by nanliu-r** [comment]
@PaulYuuu Please help to review again, thanks.

(1/1) Host_RHEL.m9.u1.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.1.0.x86_64.io-github-autotest-qemu.x86_cpu_protection_key.q35: PASS (187.31 s)

**Pull request merged**